Min # Jsonparser step5

undervineg opened the pull request.

undervine committed:
Modify 'guard' statement in containsRepeatedKey() to 'if' statement

undervine committed:
Add funcs to store linefeeds, tabs, and spaces

undervine committed:
Add func to print prettyData

undervine committed:
Add func to print prettyData

godrm commented on JSONParser/JSONParser/JSONData.swift:
```diff
@@ -35,11 +35,13 @@ struct JSONData {
         }
```
JSONData 구조체가 내부에 object와 array를 모두 갖고 있다보니 dataType 변수도 있어야 하고 
dataType에 따라서 내부 동작이 계속 반복적으로 나눠지는 것 같습니다.

JSONObject와 JSONArray를 실제로 구조체로 나누고 JSONData를 프로토콜이나 다른 형태로 추상화하는게 어땠을까요?
그러면 굳이 private(set) 으로 객체 속성을 노출하지 않았어도 될 것 같습니다.

undervine committed:
Make new structs

undervine committed:
Make struct to protocol

Bundle up common variables and functions and make them JSONData protocol.
Make extension of JSONData and implement functions that every structs conform the protocol commonly use.

undervine committed:
Separate JSONData struct to JSONArray and JSONObject structs

undervine committed:
Rename typealias of data types

undervine committed:
Reduce some functions

undervine committed:
Modify switch statment as JSONData divided into JSONArray and JSONObject struct

undervine committed:
Modify according to codacy

undervine committed:
Modify according to codacy

undervine committed:
Modify according to codacy

undervineg commented:
- JSONData 구조체를 프로토콜로 변경
- JSONObject, JSONArray 구조체 작성

godrm commented:
전체적인 구조가 바뀌었네요. 정리가 된 느낌입니다.
JSONData 프로토콜로 바뀌면서 JSONObject와 JSONArray 구조체 속성과 메소드가 모두 private가 사라졌네요.
감춰야 할 부분은 private 으로 바꿔보세요.

undervine committed:
Add private or internal keywords to variablesa and functions

undervine committed:
Add some comments

undervineg commented:
- 접근제어 키워드 추가: 프로토콜 채택함수들은 internal 키워드 추가.

godrm commented:
internal 값이 디폴트라 아무것도 안쓰면 internal 입니다. 
객체 내부에서만 쓰도록 private 를 쓰는걸 권장합니다.

undervine committed:
Try to make structs and protocol private (failed)

undervineg commented:
음.. 프로토콜에 정의된 함수나 변수들은 private을 쓸 수가 없어서 명시적으로 internal을 썼습니다. (지금은 제거)
또, 프로토콜 내부에는 ACL을 쓸 수 없어서 프로토콜(JSONData) 자체에 private을 쓰니, JSONObject나 JSONArray에서 JSONData가 보이지 않아 채택할 수가 없습니다.
다른 방법이 있을까요...?

godrm commented:
맞습니다. protocol을 만들때는 무작성 공통적인 함수가 있다고 해서 추상화를 하려고 하면 모든 메소드가 public (실제로는 internal)로 되어 버립니다. 속성도 마찬가지입니다.
따라서 protocol로 추상화를 할 때는 명확하게 그 객체를 사용하는 인터페이스가 될만한 메소드만 선언해야 합니다. 객체 내부적으로 숨겨질 구현은 private 다른 이름으로 만들어서 public 인터페이스에서 호출하도록 만들어야 합니다.

struct나 class 인터페이스가 되는 메소드는 외부에서 꼭 필요한 메소드들만 보이고, 
나머지 메소드나 속성은 private으로 감추도록 만드세요. 앞으로도 그 원칙을 지키면서 만드세요.

undervineg commented:
아하 네 알겠습니다~!

undervine committed:
Remove variables and functions from JSONData protocol definition

Remove some variables and functions want to make private.

undervineg commented:
- JSONData protocol에서 private 으로 처리할 함수, 변수 제거

godrm commented:
수고하셨습니다.
앞으로는 프로토콜로 추상화할 때는 속성보다는 메소드(함수)를 추상화하는 용도로 사용하세요.
속성을 추상화하는 개념은 클래스로 구현상속을 하는게 적합한 경우가 많습니다.
스위프트 데이터 구조 자체가 프로토콜로 속성을 추상화하고 확장하기 좋은 구조는 아닙니다. 
지금도 JSONData 에 속성만 남아버렸네요 🤔
